### PR TITLE
tone equalizer: improve pickers

### LINKED ISSUE #9995: Tone equalizer fails to update luminance mask and histogram
If a tone equalizer parameter affecting the luminance mask is changed, and is then immediately changed back to its previous setting, then the luminance mask and histogram do not update, and become out of sync with the module parameters. This happens because the luminance mask is only updated when process() is called, and it is not called if the pixelpipe cache contains a matching entry corresponding to the previously used settings.

To reproduce:
Go to the masking tab. Reset the module.
Change the mask exposure compensation and observe that the gray bar moves.
Reset the module and observe that the gray bar does not move, even though mask exposure compensation reset to 0.

Or:
Change the luminance estimator and observe that the gray bar moves.
Change the luminance estimator back to the previous setting and observe that the gray bar does not move.

I think that this explains why if you keep clicking the auto exposure compensation button repeatedly, it cycles through a few different values instead of just toggling back and forth between 0 EV and the correct value.

The problem occurs with v3.6.0 and the latest 3.7.0+1041~gb174c61be. I am testing on Arch Linux with OpenCL enabled.

### LINKED ISSUE #9997: Tone equalizer pickers inconclusive/unstable
**To Reproduce**
1. Download the raw from https://discuss.pixls.us/uploads/short-url/3x7TCchjeXvvBmtsM1eQYD1lEyo.NEF
2. Restart darktable and import
3. Enable Tone EQ
4. Try the pickers on the advanced tab several times

**Expected behavior**
Pickers pick always the same value (1st click: pick a value -- 2nd click: reset -- 3rd click: pick same value as in 1st click)

**Screencast**
tone-eq-unstable

**Which commit introduced the error**
Unknown; behaviour is the same this with 3.6.0 on Windows, no OpenCL as well as 3.7.0+1047~gf22f9ec62 on Linux with OpenCL

**Platform**
_Please fill as much information as possible in the list given below. Please state "unknown" where you do not know the answer and remove any sections that are not applicable _

* darktable version : 3.6.0 as well as 3.7.0+1047~gf22f9ec62 
* OS                : Linux kernel 5.11.0, Windows 10
* Linux - Distro    : Ubuntu 21.04
* Memory            : 16 GB (Windows), 64 GB (Linux)
* Graphics card     : none (Windows), NVidia 1060/6GB (Linux)
* CMAKE_BUILD_TYPE  : Release on Windows, RelWithDebugInfo on Linux

**Additional context**
Not always reproducible when darktable is not restarted before import.

### PR DESCRIPTION
**Pull request by Mark-64**
This PR replaces #9998 and #10014 into a clean one. 
It includes two commits:

1) Fixing erratic gui behavior of TE picker buttons (fixes #9995, fixes #9997)
corresponds to the first old PR, already reviewed by @aurelienpierre 

2) Improving the auto-exposure and auto-contrast estimations
Today they are not accurate, for two reasons:
- they are based on histogram population statistics (first and last deciles), but those are calculated on a histogram "clipped" in the range [-8, 0] EV, which is insufficient in most cases (particularly the 0 EV upper limit). The solution implemented is to use an extended histogram [-10, +6] EV for the purpose of stats calculation
- the auto-contrast estimation formula assumes that contrast correction will expand/compress the histogram simmetrically around the fulcrum in the EV space. This is true if the contrast is a log one, however TE uses a linear contrast and the former does not hold.

Therefore I implemented a new estimation math, as follows:
In the linear space (so all variable listed below are `exp2f()` of their EV counterparts) the exposure/contrast correction are applied by TE in this way

immagine

where `l` and `l'` are the luminance before and after correction, `e` is the exposure and `c` is the contrast

We want

immagine

where `l1`, `l2` are the first and last deciles, `t1`, `t2` are the target values for them (-7 EV and -1 EV). Strictly speaking, the above formulas are valid for single points, not necessarily for population stats like deciles, but the approximation is pretty good

Solving for c and e we get:

immagine

immagine

By the way I replaced deciles with "ventiles" (histogram population <5% or >95%), because the auto correction looks more accurate

### CODE REVIEW

File: src/iop/toneequal.c [hunk 1]
```diff
@@ -1725,36 +1725,33 @@ static void smoothing_callback(GtkWidget *slider, gpointer user_data)
...
     dt_dev_add_history_item(darktable.develop, self, TRUE);
-    dt_bauhaus_widget_set_quad_active(quad, FALSE);
     return;
   }
 
-  if(!g->luminance_valid || self->dev->pipe->processing)
+  if(!g->luminance_valid || self->dev->pipe->processing || !g->histogram_valid)
   {
+    ++darktable.gui->reset;
+    gboolean active = dt_bauhaus_widget_get_quad_active(quad);
```
Comment (TurboGit) [inline src/iop/toneequal.c:1752]: const

File: src/iop/toneequal.c [hunk 2]
```diff
@@ -1348,67 +1349,67 @@ static inline void build_interpolation_matrix(float A[CHANNELS * PIXEL_CHAN],
...
+  *first_decile = 16.0 * (float)first_pos / (float)(TEMP_SAMPLES - 1) - 10.0;
+  *last_decile = 16.0 * (float)last_pos / (float)(TEMP_SAMPLES - 1) - 10.0;
+
+  *max_histogram = 0;
+  // remap the extended histogram into the normal one
+  // bins between [-8; 0] EV remapped between [0 ; UI_SAMPLES]
+  for(size_t k = 0; k < TEMP_SAMPLES; ++k)
+  {
+    float EV = 16.0 * (float)k / (float)(TEMP_SAMPLES - 1) - 10.0;
+    int i = CLAMP((int)(((EV + 8.0f) / 8.0f) * (float)UI_SAMPLES), 0, UI_SAMPLES - 1);
```
Comment (TurboGit) [inline src/iop/toneequal.c:1405]: const

File: src/iop/toneequal.c [hunk 3]
```diff
@@ -1348,67 +1349,67 @@ static inline void build_interpolation_matrix(float A[CHANNELS * PIXEL_CHAN],
...
+  const int last = (int)((float)num_elem * 0.95f);
   int population = 0;
   int first_pos = 0;
   int last_pos = 0;
 
-  // scout the histogram bins looking for deciles
-  for(size_t k = 0; k < hist_bins; ++k)
+  // scout the extended histogram bins looking for deciles
+  // these would not be accurate with the regular histogram
+  for(size_t k = 0; k < TEMP_SAMPLES; ++k)
```
Comment (TurboGit) [inline src/iop/toneequal.c:1387]: Would an omp for directive will speed-up things here?

File: src/iop/toneequal.c [hunk 4]
```diff
@@ -1348,67 +1349,67 @@ static inline void build_interpolation_matrix(float A[CHANNELS * PIXEL_CHAN],
...
-  *last_decile = (hist_span * (((float)last_pos) / ((float)(hist_bins - 1)))) + hist_offset;
-}
+  // Convert decile positions to exposures
+  *first_decile = 16.0 * (float)first_pos / (float)(TEMP_SAMPLES - 1) - 10.0;
+  *last_decile = 16.0 * (float)last_pos / (float)(TEMP_SAMPLES - 1) - 10.0;
+
+  *max_histogram = 0;
+  // remap the extended histogram into the normal one
+  // bins between [-8; 0] EV remapped between [0 ; UI_SAMPLES]
+  for(size_t k = 0; k < TEMP_SAMPLES; ++k)
```
Comment (TurboGit) [inline src/iop/toneequal.c:1402]: Same as above. Would an omp for directive will speed-up things here?

File: src/iop/toneequal.c [hunk 5]
```diff
@@ -1787,53 +1798,59 @@ static void auto_adjust_exposure_boost(GtkWidget *quad, gpointer user_data)
...
     dt_dev_add_history_item(darktable.develop, self, TRUE);
-    dt_bauhaus_widget_set_quad_active(quad, FALSE);
     return;
   }
 
-  if(!g->luminance_valid || self->dev->pipe->processing)
+  if(!g->luminance_valid || self->dev->pipe->processing || !g->histogram_valid)
   {
+    ++darktable.gui->reset;
+    gboolean active = dt_bauhaus_widget_get_quad_active(quad);
```
Comment (TurboGit) [inline src/iop/toneequal.c:1825]: const

**Review by TurboGit — CHANGES_REQUESTED** [review]
Some minor changes and a possible perf opt.

Another point to raise as we are on it. The color picker in the tone equalizer are not real ones. You cannot select an area, it is kind of "auto"  tune based on the picture. What annoy me is that the color picker are left selected after clicking and doing their job. This is not good. I would at least deselect them when the adjustment is done. And we could even maybe come up with a new icon for those auto picker which do not allow for selecting an area?

File: src/iop/toneequal.c (hunk 1 above)
Comment (Mark-64) [reply src/iop/toneequal.c:1752]: Ok, I will replace the pickers with a regular non-toggle button, possibly with a new icon. Maybe a simple "A" standing for "auto", or something resembling that.

File: src/iop/toneequal.c (hunk 3 above)
Comment (Mark-64) [reply src/iop/toneequal.c:1387]: I don't think so, for simple math on an array of 256 integers (512 for the temp histogram). 
Moreover, I am not sure this cycle with an accumulator can be vectorized at all. I can slightly optimize by using two cycles, one fron the start of the array, looking for first decile and another from the end, looking for last decile, stopping the cycle when they are found. But it is very minor.

File: src/iop/toneequal.c (hunk 4 above)
Comment (Mark-64) [reply src/iop/toneequal.c:1402]: same as above. I can make a test, just for curiosity

**Comment by Mark-64** [comment]
with the last commit:
- changed the buttons icon from eyedropper to "auto levels" from retouch module, for now. If this is merged we can think of a better icon.
- the button are no longer toggle 
- minor optimization of the histogram loops. I tried to parallelize, but it is slower, for such small arrays
- added const as requested

The behavior of the button is still not fantastic because bauhaus buttons next to sliders are actually made for being toggle. When set as no toggle they do not provide any feedback on click to the user, and no prelight either. They almost seem inactive.
I don't know if there is a way to place a normal button in the same place.
Moreover, with the previous implementation, the toggle button had the subtle function to reset the correction and recalculate the histogram and the relative deciles, which means that a new auto-exposure/contrast calculation will always restart from the same data.
Without the toggle, the previous correction must be reverted "in theory" by the auto-correction math, which is what is done, but it's not perfect.
This is because the effect of the surface blur on the histogram is a bit unpredictable, especially for eigf (see also #10034) so it cannot be reverted accurately without actually recalculate the histogram.
The bottom line effect is that clicking mutliple time on the buttons lead to slighlty different estimations, especially when a contrast is dialed in.

File: src/iop/toneequal.c (hunk 4 above)
Comment (aurelienpierre) [reply src/iop/toneequal.c:1415]: OpenMP can't be used here, it would create a race condition. We could use `#pragma omp for reduction(+:histogram)` to avoid it if `histogram` was a scalar, but here it's an array and OpenMP didn't implement reductions on arrays/vector last time I checked.

**Comment by Mark-64** [comment]
small modification in bauhaus.c to highlight the quad button when clicked in non-toggle mode, in order to provide visual feedback to the user

**Comment by TurboGit** [comment]
Ok, to me this is good to go. I'm not fond of the icon at all but this can be worked later when we'll get more feedback. I'm not in a position to propose a better icon for now anyway.
